File: codes/python/685.冗余连接-ii.py
class UnionFind:
    def __init__(self, n: int):
        self.parent = [i for i in range(n)]
        self.rank = [1 for i in range(n)]

    def find(self, x: int) -> int:
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX == rootY:
            return False
        if self.rank[rootX] < self.rank[rootY]:
            rootX, rootY = rootY, rootX
        self.parent[rootY] = rootX
        self.rank[rootX] += self.rank[rootY]
        return True

class Solution:
    def findRedundantDirectedConnection(self, edges: List[List[int]]) -> List[int]:
        ans = [0, 0]
        inDegress = self.getNodeInDegress(edges)
        for i in range(len(edges) - 1, -1, -1):
            if inDegress[edges[i][1]] == 2:
                if self.isTree(edges, i):
                    ans[0] = edges[i][0]
                    ans[1] = edges[i][1]
                    return ans
        unionFind = UnionFind(len(edges) + 1)
        for edge in edges:
            if not unionFind.union(edge[0], edge[1]):
                ans[0] = edge[0]
                ans[1] = edge[1]
                return ans
        return ans
    
    def isTree(self, edges: List[List[int]], removeIndex: int) -> bool:
        unionFind = UnionFind(len(edges) + 1)
        for i in range(len(edges)):
            if i == removeIndex:
                continue
            if not unionFind.union(edges[i][0], edges[i][1]):
                return False
        return True
    
    def getNodeInDegress(self, edges: List[List[int]]) -> List[int]:
        inDegrees = [0 for i in range(len(edges) + 1)]
        for edge in edges:
            inDegrees[edge[1]] += 1
        return inDegrees

File: codes/java/685.冗余连接-ii.java

class UnionFind {
	private int[] parent;
	private int[] rank;

	public UnionFind(int n) {
		parent = new int[n];
		rank = new int[n];
		for (int i = 0; i < n; i++) {
			parent[i] = i;
			rank[i] = 1;
		}
	}

	public int find(int x) {
		if (x != parent[x]) {
			parent[x] = find(parent[x]);
		}
		return parent[x];
	}

	public boolean union(int x, int y) {
		int rootX = find(x);
		int rootY = find(y);
		if (rootX == rootY) {
			return false;
		}
		if (rank[rootX] < rank[rootY]) {
			int temp = rootX;
			rootX = rootY;
			rootY = temp;
		}
		parent[rootY] = rootX;
		rank[rootX] += rank[rootY];
		return true;
	}
}

class Solution {
	public int[] findRedundantDirectedConnection(int[][] edges) {
		int[] ans = new int[2];
		// 获取每个节点的入度
		int[] inDegress = getNodeInDegress(edges);
		// 首先判断第一种情况,从后往前遍历,寻找是否存在入度为2的顶点,如果存在,则判断是否为树,如果是树,则返回,否则继续
		for (int i = edges.length - 1; i >= 0; i--) {
			if (inDegress[edges[i][1]] == 2) {
				if (isTree(edges, i)) {
					ans[0] = edges[i][0];
					ans[1] = edges[i][1];
					return ans;
				}
			}
		}
		// 接着判断第二种情况，每个节点的入度都为1，判断是否成环
		UnionFind unionFind = new UnionFind(edges.length + 1);
		for (int[] edge : edges) {
			if (!unionFind.union(edge[0], edge[1])) {
				ans[0] = edge[0];
				ans[1] = edge[1];
				return ans;
			}
		}
		return ans;
	}

	private boolean isTree(int[][] edges, int removeIndex) {
		UnionFind unionFind = new UnionFind(edges.length + 1);
		for (int i = 0; i < edges.length; i++) {
			if (i == removeIndex) {
				continue;
			}
			if (!unionFind.union(edges[i][0], edges[i][1])) {
				return false;
			}
		}
		return true;
	}

	private int[] getNodeInDegress(int[][] edges) {
		int[] inDegrees = new int[edges.length + 1];
		for (int[] edge : edges) {
			inDegrees[edge[1]]++;
		}
		return inDegrees;
	}
}